ports/zephyr/zephyr_device.c

```diff
@@ -27,6 +27,81 @@
 #include "zephyr_device.h"
 #include "py/runtime.h"
 
+#if defined(CONFIG_MICROPY_DYNAMIC_PINCTRL)
+
+#include <zephyr/drivers/pinctrl.h>
+extern pinctrl_soc_pin_t zephyr_pinctrl_get_pinctrl(const mp_obj_t obj);
+
+/* Circumvent limitation of only having devices that are initialized */
+typedef struct mp_zephyr_device_data_t_ {
+    const struct device *dev;
+    char *name;
+    const char **labels;
+    size_t label_cnt;
+    struct pinctrl_dev_config *pinctrl_cfg;
+} mp_zephyr_device_data_t;
+
+#define PINCTRL_EXISTS_X_OR_Y(node, to_do, to_or) COND_CODE_1(DT_CAT(node, _PINCTRL_IDX_0_EXISTS), \
+    (to_do), (to_or))
+
+#define DT_NODELABEL_STRING_ARRAY_DECLARE(node) static const char *_##node##_labels[] = \
+    DT_NODELABEL_STRING_ARRAY(node)
+
+#define DT_NODELABEL_STRING_ARRAY_DECLARE_PINCTRL(node) \
+    PINCTRL_EXISTS_X_OR_Y(node, DT_NODELABEL_STRING_ARRAY_DECLARE(node); , )
+
+DT_FOREACH_STATUS_OKAY_NODE(DT_NODELABEL_STRING_ARRAY_DECLARE_PINCTRL)
+
+#define DT_NODELABEL_STRING_ARRAY_NAME(node) _##node##_labels
+
+#define PINCTRL_DT_DEV_CONFIG_ENTRY(node) \
+    PINCTRL_EXISTS_X_OR_Y(node, PINCTRL_DT_DEV_CONFIG_GET(node), NULL)
+
+#define PINCTRL_DT_DATA_DECLARE(node) \
+    PINCTRL_EXISTS_X_OR_Y(node, PINCTRL_DT_DEV_CONFIG_DECLARE(node); , )
+
+DT_FOREACH_STATUS_OKAY_NODE(PINCTRL_DT_DATA_DECLARE)
+
+#define DEVICE_DATA_ENTRY(node) {                        \
+        .dev = DEVICE_DT_GET(node),                      \
+        .name = DEVICE_DT_NAME(node),                    \
+        .labels = DT_NODELABEL_STRING_ARRAY_NAME(node),  \
+        .label_cnt = DT_CAT(node, _NODELABEL_NUM),       \
+        .pinctrl_cfg = PINCTRL_DT_DEV_CONFIG_ENTRY(node) \
+},
+
+#define DEVICE_DATA_ENTRY_PINCTRL(node) \
+    PINCTRL_EXISTS_X_OR_Y(node, DEVICE_DATA_ENTRY(node), )
+
+static const mp_zephyr_device_data_t all_devices[] =
+{
+    DT_FOREACH_STATUS_OKAY_NODE(DEVICE_DATA_ENTRY_PINCTRL)
+};
```

Comment:
For MicroPython maintainers: what I understood from the macros:

With `CONFIG_MICROPY_DYNAMIC_PINCTRL=y`, only peripherals with `deferred-init` will be usable 

- `all_devices[]` > `device_find_by_name/nodelabel()` > `zephyr_device_setup()` > `make_new()`

It is already required to have `status = "okay";` added for every peripheral to be enabled, so adding `zephyr,deferred-init;`, as needed by these macros, is just a 1 line addition:

```diff
 &i2c0 {
 	status = "okay";
+	zephyr,deferred-init;
 };
```

The rest of the config such as `pinctrl-0 = <&uart1_default>;` block would have been needed to enable `uart1` anyway.

So the current solution seems like smallest amount of manual work possible, going below 1 line would require change on upstream Zephyr.

Reply:
Or peripherals that support de-init, which is 2 whole nrf peripherals atm :)

I intend to add de-init to bflb at least so that will double the number!

Reply:
Cutting edge Zephyr support I see! :)